### ISSUE TITLE
Remove dependency on distribute

### ISSUE THREAD

**Issue by merwok** [zatosource/zato-apitest#24]
distribute is now merged bad into setuptools, which is actively maintained.  In a recent virtualenv with up-to-date setuptools and pip, there is an exception when trying to install zato-apitest.

**Comment by dsuch** [comment]
@merwok - if you can advise on what is needed, please do.

I don't keep track of what is currently actively maintained on that front - personally I prefer zc.buildout which has been a stable choice for years now.

Thanks a lot.

**Comment by merwok** [comment]
buildout, like pip, is the class high-level package manager.  setuptools (or its fork distribute) is the lower-level library that is used by both.  If you don’t import setuptools or pkg_resources at run time and depend on a specific version or their API, then you could just remove distribute from your requirements.

**Comment by dsuch** [comment]
Hi @merwok - yes, I understand which is which.

I'm just saying that I am content with zc.buildout which is stable and lets me focus on my own job instead of making me go back to obscure low-level packaging issues several times a year.

pkg_resources happens to be used like here:

```python
__import__('pkg_resources').declare_namespace('zato')
```

https://github.com/zatosource/zato-apitest/blob/master/src/zato/__init__.py

I'm not really sure if this is still a recommended declaration and if it's required or accepted by any tools nowadays. Can it be safely removed?

If distribute is not to be used anymore, let's do away with it, I'm all for it - but please, help me out here and submit a PR, I'm just not up to date on how to tackle such a migration.

Thanks again.

**Comment by merwok** [comment]
The namespace package declaration should not be removed.

The fix I’m proposing is to replace distribute with setuptools in your dependencies (without version specifier).  I’ll probably be able to make a PR tomorrow.

**Comment by dsuch** [comment]
> without version specifier

That won't work @merwok - there cannot be any dependencies that are unspecified. This is a strict principle in all Zato-related projects.

All builds must always be reproducible and each version must always be exact. setuptools is just a dependency, there is nothing magical about it, and I'm not going to take any chances.

I understand the reason you need it is because you have a dependency on zato-apitest in your own virtualenv and in your dependencies you don't pin them?

How do virtualenv/setuptools deal with such situations elsewhere?

**Comment by dsuch** [comment]
By the way @merwok - the reason this has never been caught before is that as far as I can tell zato-apitest is typically used either standalone or under Jenkins.

On top of that, when you install the tool from source .. 

```
$ git clone https://github.com/zatosource/zato-apitest.git && cd zato-apitest
$ make
```

.. it already installs it under virtualenv.

So maybe this suffices?

**Comment by dsuch** [comment]
Yet another one - how about you set the dependency on setuptools in zato-apitest to one that is currently the latest one?

If with time you'd like to change it, just let me know by sending a PR, and it can be upped?

**Comment by SpotlightKid** [comment]
I was just about to file an issue about the same thing.

The dependency on distribute prevents zato-apitest from being installed via "pip install zato-apitest" in a virtual environment.

Installation fails on trying to install distribute with the following error:

```
Collecting distribute==0.6.34 (from zato-apitest)
  Downloading distribute-0.6.34.tar.gz (644kB)
    100% |################################| 647kB 243kB/s 
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/home/chris/tmp/pip-build-iY5c4D/distribute/setup.py", line 248, in <module>
        scripts = scripts,
      File "/usr/lib64/python2.7/distutils/core.py", line 111, in setup
        _setup_distribution = dist = klass(attrs)
      File "setuptools/dist.py", line 225, in __init__
        _Distribution.__init__(self,attrs)
      File "/usr/lib64/python2.7/distutils/dist.py", line 287, in __init__
        self.finalize_options()
      File "setuptools/dist.py", line 257, in finalize_options
        ep.require(installer=self.fetch_build_egg)
      File "pkg_resources.py", line 2025, in require
        working_set.resolve(self.dist.requires(self.extras),env,installer))
      File "pkg_resources.py", line 2235, in requires
        dm = self._dep_map
      File "pkg_resources.py", line 2464, in _dep_map
        self.__dep_map = self._compute_dependencies()
      File "pkg_resources.py", line 2497, in _compute_dependencies
        common = frozenset(reqs_for_extra(None))
      File "pkg_resources.py", line 2494, in reqs_for_extra
        if req.marker_fn(override={'extra':extra}):
      File "_markerlib/markers.py", line 109, in marker_fn
        return eval(compiled_marker, environment)
      File "<environment marker>", line 1, in <module>
    NameError: name 'sys_platform' is not defined
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
    
      File "<string>", line 20, in <module>
    
      File "/home/chris/tmp/pip-build-iY5c4D/distribute/setup.py", line 248, in <module>
    
        scripts = scripts,
    
      File "/usr/lib64/python2.7/distutils/core.py", line 111, in setup
    
        _setup_distribution = dist = klass(attrs)
    
      File "setuptools/dist.py", line 225, in __init__
    
        _Distribution.__init__(self,attrs)
    
      File "/usr/lib64/python2.7/distutils/dist.py", line 287, in __init__
    
        self.finalize_options()
    
      File "setuptools/dist.py", line 257, in finalize_options
    
        ep.require(installer=self.fetch_build_egg)
    
      File "pkg_resources.py", line 2025, in require
    
        working_set.resolve(self.dist.requires(self.extras),env,installer))
    
      File "pkg_resources.py", line 2235, in requires
    
        dm = self._dep_map
    
      File "pkg_resources.py", line 2464, in _dep_map
    
        self.__dep_map = self._compute_dependencies()
    
      File "pkg_resources.py", line 2497, in _compute_dependencies
    
        common = frozenset(reqs_for_extra(None))
    
      File "pkg_resources.py", line 2494, in reqs_for_extra
    
        if req.marker_fn(override={'extra':extra}):
    
      File "_markerlib/markers.py", line 109, in marker_fn
    
        return eval(compiled_marker, environment)
    
      File "<environment marker>", line 1, in <module>
    
    NameError: name 'sys_platform' is not defined
    
    ----------------------------------------
    Command "python setup.py egg_info" failed with error code 1 in /home/chris/tmp/pip-build-iY5c4D/distribute
```

**Comment by dsuch** [comment]
Thanks @SpotlightKid - let's remove it then, it's all good by me if it doesn't break anything and only helps :-)

**Comment by SpotlightKid** [comment]
I don't know how buildout works, but installing via pip or easy_install or "python setup.py install" invokes "setup.py" in all cases, which imports setuptools, so implicitly already pre-requires it.

Even if you install all requirements with "pip install -e requirements.txt", you already need setuptools to install pip. if you want to include `setuptools` in `requirements.txt`, maybe it should be filtered out in  `parse_requirements` in `setup.py`, so you don't have to pin it's version.

And if you always pin versions, how do you handle versions that disappear from PyPI? zato-common currently suffers from that problem (bzr 2.2.3 isn't available on PyPI anymore).

BTW, `parse_requirements` isn't very robust, since the format of the file allows to include lines like 

    -r requirements-common.txt

**Comment by dsuch** [comment]
@SpotlightKid - I am all for removing distribute if it's not what one should use these days, you don't need to convince me.

I'm just saying that I'd like to pin the new dependency so that the builds are reproducible and everyone gets the same result in the end without 'works here but not there yet over there still it does' symptoms.

> how do you handle versions that disappear from PyPI?

There's no magic, unfortunately. The only thing that helps is one's own private package index which, given how lightly people treat deleting older releases from PyPI, is something I'd recommend anyway.

So to recap - I am very interested in making zato-apitest as accessible as it can be done but there are some conventions, like version pinning, that I'd like to keep.

Thanks.

**Comment by SpotlightKid** [comment]
@dsuch My argument was more for not including `setuptools` in the requirements in `setup.py`, whether with a pinned version or not and independently of whether to list it in requirements.txt or not.

**Comment by dsuch** [comment]
Ok. But seeing as it gets into setup.py through requirements.txt you're basically saying not to depend on it at all?

https://github.com/zatosource/zato-apitest/blob/master/setup.py#L77

**Comment by SpotlightKid** [comment]
No, if you want to have all dependencies listed in `requirements.txt`, that's another thing. But having setuptools as a dependency in `setup.py` interferes with virtual environments.

Like I said, maybe you could just filter ` setuptoools` out in `parse_requirements`.

**Comment by dsuch** [comment]
Cool, I see @SpotlightKid . So the plan so far is:

* Remove distribute
* Add setuptools to requirements.txt
* Make sure setuptools doesn't make it to setup.py

And the only thing to confirm with @merwok is whether it's OK to have a pinned version of setuptools in requirements.txt

Thank you both for the input so far!

**Comment by merwok** [comment]
Pinning to the latest working version of setuptools would be fine.  If I need a newer version later, I’ll just propose a pull request here.  Thanks!

**Comment by merwok** [comment]
Here’s the pull request.

**Comment by dsuch** [comment]
@merwok and @SpotlightKid,

version 1.9 has just been released - it contains a dependency on setuptools but only in requirements.txt.

Per @SpotlightKid's suggestion the dependency is not carried over to setup.py's install_requires.

Everything installs cleanly under virtualenv, tests run fine and demo works correctly.

Thanks again for opening the ticket and helping out with everything.
